presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java

```diff
@@ -0,0 +1,223 @@
+ * simplify this.
+ * i.e, ExpressionInterpreter can discard all conditions in which result != value
+ */
+public class RewriteCaseExpressionPredicate
+        implements Rule<FilterNode>
+{
+    private static final Pattern<FilterNode> PATTERN = filter();
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    public RewriteCaseExpressionPredicate(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionManager is null");
+    }
+
+    private static class Rewriter
+            extends RowExpressionRewriter<Void>
+    {
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final FunctionResolution functionResolution;
+        private final LogicalRowExpressions logicalRowExpressions;
+
+        private Rewriter(FunctionAndTypeManager functionAndTypeManager)
+        {
+            this.functionAndTypeManager = functionAndTypeManager;
+            this.functionResolution = new FunctionResolution(functionAndTypeManager);
+            this.logicalRowExpressions = new LogicalRowExpressions(new RowExpressionDeterminismEvaluator(this.functionAndTypeManager),
+                    this.functionResolution,
+                    this.functionAndTypeManager);
+        }
+
+        @Override
+        public RowExpression rewriteCall(CallExpression node, Void context, RowExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (functionResolution.isComparisonFunction(node.getFunctionHandle())) {
+                if (node.getArguments().size() == 2) {
+                    RowExpression left = node.getArguments().get(0);
+                    RowExpression right = node.getArguments().get(1);
+                    if (isCaseExpression(left)) {
+                        return processCaseExpression(new SwitchExpression((SpecialFormExpression) left), right, node.getFunctionHandle(), node.getDisplayName(), false);
+                    }
+                    else if (isCaseExpression(right)) {
+                        return processCaseExpression(new SwitchExpression((SpecialFormExpression) right), left, node.getFunctionHandle(), node.getDisplayName(), true);
+                    }
+                }
+            }
+            return null;
+        }
+
+        private boolean isCaseExpression(RowExpression expression)
+        {
+            return expression instanceof SpecialFormExpression && ((SpecialFormExpression) expression).getForm().equals(SWITCH);
+        }
+
+        private RowExpression processCaseExpression(SwitchExpression switchExpression,
+                                                    RowExpression otherExpression,
+                                                    FunctionHandle functionHandle,
+                                                    String displayName,
+                                                    boolean switchExpressionOnRight)
+        {
```

Comment:
We have scenarios such as 
```
(CASE
    WHEN (replace(column_name, 'prefix_string')) = 'dept1' THEN 'V1'
    WHEN (replace(column_name, 'prefix_string')) = 'dept2'  THEN 'V2
    ELSE (replace(column_name, 'prefix_string'))
    END
) = UPPER('v1');
```
In this case checking if it is just Constant Expression would be problematic.

Reply:
One problem we see with CASE expression is that it doesn't extract out the Column Domain while picking table layout. Switching it to AND/OR cases helps achieve it. Even if there is a column reference present in the equals side 
```
(CASE
    WHEN (replace(column_name, 'prefix_string')) = 'dept1' THEN 'V1'
    WHEN (replace(column_name, 'prefix_string')) = 'dept2'  THEN 'V2
    ELSE 'V3'
    END
) = column_name_2;
```
Above expression gets converted into a cumbersome AND/OR expression since column value's equality with THEN value cannot be evaluated and the expression won't be further simplified, but still it helps RowExpressionInterpretter/Optimizers in extracting column domain for the table from that expression. i.e, column_name_2 domain is set as [ V1, V2, V3].

Reply:
These things can lead to subtle bugs. And also if the other side is a function call pushing it into the CASE can make it evaluate many times. We should not need to do random stuff like upper(..). But maybe extend it to just constant or a variable/field reference/input expression.